sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,668 @@
+    return (_fp_group, _fp_group._generators)
+
+@public
+def vfp_group(fr_grpm, relators):
+    _fp_group = FpGroup(symbols, relators)
+    pollute([sym.name for sym in _fp_group.symbols], _fp_group.generators)
+    return _fp_group
+
+
+def _parse_relators(rels):
+    """Parse the passed relators."""
+    return rels
+
+_fp_group_cache = {}
+
+class FpGroup(Basic):
+    """
+    The FpGroup would take a FreeGroup and a list/tuple of relators, the
+    relators would be specified in such a way that each of them be equal to the
+    identity of the provided free group.
+    """
+    is_FpGroup = True
+
+    def __new__(cls, fr_grp, relators):
+        relators = _parse_relators(relators)
+        # return the corresponding FreeGroup if no relators
+        # is specified
+        if not relators:
+            return fr_grp
+        obj = Basic.__new__(cls, fr_grp, relators)
+        obj._free_group = fr_grp
+        obj._relators = relators
+        obj.generators = obj._generators()
+        obj.dytype = type("FpGroupElm", (FpGroupElm,), {"group": obj})
+        return obj
+
+    @property
+    def free_group(self):
+        return self._free_group
+
+    def relators(self):
+        return tuple(self._relators)
+
+    def _generators(self):
+        """Returns the generators of the associated free group."""
+        return self.free_group.generators
+
+    def subgroup(self, gens):
+        # gens is list of generators of self.
+        pass
+
+    def coset_table(self, Y):
+        R = self.relators
+
+
+# sets the upper limit on the number of cosets generated during
+# Coset Enumeration. "M" from Derek Holt's. It is supposed to be
+# user definable.
+DefaultMaxLimit = 500
+
+# fp_grp: Finitely Presented Group with < X|R > as presentation.
+# H: subgroup of fp_grp.
+# NOTE: We start with H as being only a list of words in generators
+#       of "fp_grp". Since `.subgroup` method has not been implemented.
+
+def coset_table(fp_grp, H):
+    # list of generators and their inverses, should be made a tuple.
+    A = []
+    for gen in fp_grp:
+        A.extend([gen, gen.inverse()])
+    r = len(A)
+    # start with table of 1 row (can change) and 2*(no. of gens) coloumns (fixed)
+    table = [r*[None]]
+
+# coset_table: Mathematically a coset table
+#               represented using a list of lists
+# alpha: Mathematically a coset (precisely, a live coset)
+#       represented by an integer between i with 1 <= i <= n
+#       α ∈ c
+# x: Mathematically an element of "A" (set of generators and
+#   their inverses), represented using "FpGroupElm"
+
+# Pg. 153
+## later: remove "A" from argument
+def define(C, alpha, x, A):
+    # now C.n is not obtainable
+    if len(C) == DefaultMaxLimit:
+        # abort the further generation of cosets
+        return
+    C.append([None]*len(A))
+    # beta is the new coset generated
+    beta = len(C) - 1
+    C[alpha][A.index(x)] = beta
+    C[beta][A.index(x.inverse())] = alpha
+
+p = []
```

Comment:
I think this is logically a part of the table together with the action data. We should probably have `class CosetTable`.